### PR TITLE
Add ament_doxygen_generate() macro

### PR DESCRIPTION
**Pull request by hidmic**
Precisely what the title says.

### CODE REVIEW

File: package.xml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="2">
+  <name>ament_cmake_doxygen</name>
+  <version>0.6.2</version>
+  <description>CMake macros to help generating documentation with doxygen</description>
+  <maintainer email="daniel.stonier@tri.global">Daniel Stonier</maintainer>
+  <license>TODO</license>
```
Comment (agalbachicar) [inline package.xml]: BSD?

Comment (agalbachicar) [reply package.xml]: I think this is mostly to you, @stonier .

File: cmake/ament_doxygen_generate.cmake [hunk 2]
```diff
@@ -0,0 +1,74 @@
+
```
Comment (clalancette) [inline cmake/ament_doxygen_generate.cmake:14]: Going out on a bit of a limb here, but I think we should add the license header for Apache 2.0 with a copyright holder of Toyota Research Institute here.  @stonier can you confirm?

File: package.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="2">
+  <name>ament_cmake_doxygen</name>
+  <version>0.6.2</version>
```
Comment (clalancette) [inline package.xml]: Since this is a new package in a new repository, I'd suggest version `0.1.0`.

File: package.xml (hunk 1 above)
Comment (clalancette) [inline package.xml]: Again, need confirmation from @stonier, but probably Apache 2.0.

File: resources/Doxyfile.in [hunk 4]
```diff
@@ -0,0 +1,344 @@
+# Doxyfile 1.8.15
```
Comment (clalancette) [inline resources/Doxyfile.in:1]: As a "generic" utility, this configuration is pretty opinionated on how it lays out Doxygen.  I guess we can always add configurability later, so I'm OK with this for now.

File: resources/doxygen_extra.css [hunk 5]
```diff
@@ -0,0 +1,101 @@
+/* s3://drake-web/fonts/stylesheet.css */
+@import url('https://dmn3132s4wr9m.cloudfront.net/fonts/stylesheet.css');
```
Comment (agalbachicar) [inline resources/doxygen_extra.css]: Can we make a copy an add the proper license? I would prefer to be as autonomous as possible with this.

**Review by clalancette — COMMENTED** [review]
Mostly comments about the license and version numbers.  Looks good to me otherwise.

File: resources/Doxyfile.in [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+SHOW_GROUPED_MEMB_INC  = NO
+FORCE_LOCAL_INCLUDES   = NO
+INLINE_INFO            = NO
+SORT_MEMBER_DOCS       = YES
+SORT_BRIEF_DOCS        = NO
+SORT_MEMBERS_CTORS_1ST = NO
+SORT_GROUP_NAMES       = NO
+SORT_BY_SCOPE_NAME     = NO
+STRICT_PROTO_MATCHING  = NO
+GENERATE_TODOLIST      = NO
```
Comment (agalbachicar) [inline resources/Doxyfile.in]: I think we would benefit from having this.

File: resources/Doxyfile.in [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+                         *.h++
+RECURSIVE              = YES
+EXCLUDE                =
+EXCLUDE_SYMLINKS       = NO
+EXCLUDE_PATTERNS       =
+EXCLUDE_SYMBOLS        = 
+EXAMPLE_PATH           =
+EXAMPLE_PATTERNS       =
+EXAMPLE_RECURSIVE      = NO
+IMAGE_PATH             =
```
Comment (agalbachicar) [inline resources/Doxyfile.in:137]: Nit: @stonier you might want to add TRI's logo?

File: cmake/ament_doxygen_generate.cmake [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  endforeach()
+  set(GENERATE_TAGFILE "${OUTPUT_DIRECTORY}/${PROJECT_NAME}.tag")
+
+  configure_file("${CONFIG_DIRECTORY}/Doxyfile.in" "${BUILD_DIRECTORY}/Doxyfile" @ONLY)
+
+  if (args_CONFIG_OVERLAY)
+    string_ends_with("${args_CONFIG_OVERLAY}" ".in" is_template)
+    if(is_template)
+      get_filename_component(overlay_filename "${args_CONFIG_OVERLAY}" NAME)
+      # cut of .in extension
```
Comment (agalbachicar) [inline cmake/ament_doxygen_generate.cmake]: Isn't it `cut off`?

**Review by agalbachicar — CHANGES_REQUESTED** [review]
I left some comments and asked a question, similar to @clalancette regarding the license. Not a strong opinion on which one to use.

File: resources/Doxyfile.in (hunk 4 above)
Comment (hidmic) [reply resources/Doxyfile.in:1]: That's what `CONFIG_OVERLAY` is meant for. The idea was to provide reasonable defaults to reduce duplication, but downstream packages can always override all of them. Including the ones that make this work, so it should probably be documented what must be preserved.

File: cmake/ament_doxygen_generate.cmake (hunk 2 above)
Comment (stonier) [reply cmake/ament_doxygen_generate.cmake:14]: Copyright Toyota Research Institute / License Apache 2.0 is fine by me.

File: package.xml (hunk 1 above)
Comment (stonier) [reply package.xml]: Apache 2.0 to match ament_cmake's license works for me.

Comment (stonier) [reply package.xml]: Confirmed.

File: package.xml (hunk 3 above)
Comment (stonier) [reply package.xml]: +1

File: resources/doxygen_extra.css (hunk 5 above)
Comment (hidmic) [reply resources/doxygen_extra.css]: So, that's trickier than it seems. It's not just a stylesheet, but all the fonts too. I guess I can try replicate them all and rewrite SASS to hit those instead.

Comment (hidmic) [reply resources/doxygen_extra.css]: Alternatively, we can drop these custom fonts :)

Comment (agalbachicar) [reply resources/doxygen_extra.css]: +1 to drop.

File: cmake/ament_doxygen_generate.cmake (hunk 8 above)
Comment (hidmic) [reply cmake/ament_doxygen_generate.cmake]: Corrected in https://github.com/ToyotaResearchInstitute/ament_cmake_doxygen/pull/1/commits/5ce4b79b6ccae7c8be08097f817f22d585f1e0bf

File: resources/Doxyfile.in (hunk 6 above)
Comment (hidmic) [reply resources/Doxyfile.in]: Flipped the switch in https://github.com/ToyotaResearchInstitute/ament_cmake_doxygen/pull/1/commits/5ce4b79b6ccae7c8be08097f817f22d585f1e0bf

File: resources/doxygen_extra.css (hunk 5 above)
Comment (hidmic) [reply resources/doxygen_extra.css]: Dropping the entire stylesheet (it was only used to add these fonts) in https://github.com/ToyotaResearchInstitute/ament_cmake_doxygen/pull/1/commits/5ce4b79b6ccae7c8be08097f817f22d585f1e0bf

**Review by clalancette — APPROVED** [review]

**Review by agalbachicar — APPROVED** [review]
LGTM

**Pull request merged**